### PR TITLE
Improved error message in Rebin when sentinel values not set

### LINKED ISSUE #20000: Rebin fails on Exercise 4 of Introduction to Python in Mantid (Extending Python With Mantid)
The current automated solution - that I've written. Rebin seems to fail when monitors are included - succeeds by excluding monitors. Tested manually and achieved same outcomes.

``` python
from mantid.kernel import *
from mantid.api import *
from mantid.simpleapi import Load, ConvertUnits, Rebin
from numpy import *

class ConvertToEnergy(PythonAlgorithm):

    def PyInit(self):
        self.declareProperty(FileProperty(name="Filename",
                                          defaultValue="",
                                          action=FileAction.Load))
        self.declareProperty(name="Ei", defaultValue=0.0,
                             validator=FloatMandatoryValidator())
        self.declareProperty(FloatArrayProperty(name="BinParams",
                                                validator=FloatArrayLengthValidator(3),
                                                direction=Direction.Input))
        self.declareProperty(WorkspaceProperty(name="OutputWorkspace",
                                               defaultValue="",
                                               direction=Direction.Output))

    def PyExec(self):
        _tmpws = Load(Filename=self.getPropertyValue("Filename")) # Excluding Monitors fixes this
        _tmpws = ConvertUnits(InputWorkspace=_tmpws, Target="DeltaE",
                              EMode="Direct", EFixed=self.getPropertyValue("Ei"))
        _tmpws = Rebin(InputWorkspace=_tmpws, Params=self.getPropertyValue("BinParams"))
        output = WorkspaceFactory.create(_tmpws, NVectors=1)
        output.setX(0, _tmpws.dataX(0))
        spectraSum = np.zeros(_tmpws.getNumberBins())

        for i in range(_tmpws.getNumberBins()):
            sum = 0

            for j in range(_tmpws.getNumberHistograms()):
                sum += _tmpws.dataY(j)[i]

            spectraSum[i] = sum

        _tmpws.delete()
        output.setY(0, spectraSum)
        self.setProperty("OutputWorkspace", OutputWorkspace)

AlgorithmFactory.subscribe(ConvertToEnergy)
```

### PR DESCRIPTION
**Pull request by josephframsay**
**Description of work.**

It was found that, when rebinning a workspace containing monitors after running ConvertUnits on that workspace, an InvalidBinEdgesError was thrown. This is due to cells in the monitor being assigned to `DBL_MIN` as no sensible value could be produced. No more elegant fix could be found, but a more meaningful error message is now given when the above occurs.

**To test:**
```
from mantid.kernel import *
from mantid.api import *
from mantid.simpleapi import Load, ConvertUnits, Rebin
from numpy import *

_tmpws = LoadRaw(Filename="MAR11015.raw", LoadMonitors="Include") # Excluding Monitors fixes this
_tmpws = ConvertUnits(InputWorkspace=_tmpws, Target="DeltaE",
                      EMode="Direct", EFixed="12.9729")
_tmpws = Rebin(InputWorkspace=_tmpws, Params="-11,0.01,11")
```

Running this script used to produce `Negative or zero bin widths not allowed.` - not very meaningful for the user in this case. It should now output `One or more x-values was unusually low (below -1e100). This usually occurs when a monitor spectrum has not been masked after ConvertUnits has been run on the workspace`.

Fixes #20000  

**Release Notes** 
*Does not need to be in the release notes.*

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [x] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/HistogramData/src/Rebin.cpp [hunk 1]
```diff
@@ -40,6 +40,13 @@ Histogram rebinCounts(const Histogram &input, const BinEdges &binEdges) {
     auto owidth = xo_high - xo_low;
     auto nwidth = xn_high - xn_low;
 
+    if (xo_high < -1e100 || xo_low < -1e100) {
```
Comment (AntonPiccardoSelg) [inline Framework/HistogramData/src/Rebin.cpp]: In this case is either `xo_high` or `xo_low` equal `DBL_MIN`? If so is there are reason not to `if (xo_high == DBL_MIN || xo_low < DBL_MIN)`?

**Review by AntonPiccardoSelg — CHANGES_REQUESTED** [review]

File: Framework/HistogramData/src/Rebin.cpp (hunk 1 above)
Comment (josephframsay) [inline Framework/HistogramData/src/Rebin.cpp]: Neither would equal DBL_MIN - they come out as -DBL_MAX if the error I described occurs. The problem originates in ConvertUnits::getDetectorValues where, for want of a better value, efixed is set to DBL_MIN. The error then propagates through to here. I've made the check more explicit.

**Review by AntonPiccardoSelg — APPROVED** [review]
Checked the error message when the test script is run and it looks good to me.

**Review by SimonHeybrock — CHANGES_REQUESTED** [review]
You are adding code to the inner loop in a low level function.
- Have you checked the performance tests? This might be influencing all users of `Rebin`.
- I am not sure that this extra check will not make valid use fail. In particular `xo_low == - DBL_MAX` looks like something that may occur?
- I am not happy with referencing higher-level concepts such as monitors and the `ConvertUnits` algorithm in a low level library that is not aware of those concepts. Is there maybe another solution?

**Comment by josephframsay** [comment]
There's no noticeable effect on performance that we could find. Could a compromise be to make this a special case of the previously existing error check, ie to look for it only once we know that `owidth == 0`? This would have even less effect on performance, and would only fail in the case of real failures. Though I see what you mean about referencing higher-level concepts in a low-level library, the problem only bubbles to the surface at this point, as it's hidden in the internals of other classes until then.

**Comment by SimonHeybrock** [comment]
Yes that sounds better.

**Review by SimonHeybrock — APPROVED** [review]

**Pull request merged**